File: usb-cellular-tether/android/app/src/main/java/com/example/usbcellulartether/FullTunnelProtocol.kt
package com.example.usbcellulartether

import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer

enum class FullTunnelMessageType(val code: Int) {
    OpenTcp(1),
    TcpData(2),
    CloseTcp(3),
    UdpData(4),
    Error(5);

    companion object {
        fun fromCode(code: Int): FullTunnelMessageType =
            entries.firstOrNull { it.code == code } ?: error("Unknown message type: $code")
    }
}

data class FullTunnelMessage(
    val type: FullTunnelMessageType,
    val sessionId: Int = 0,
    val host: String? = null,
    val port: Int = 0,
    val data: ByteArray = ByteArray(0),
)

class FullTunnelProtocol(private val input: InputStream, private val output: OutputStream) {
    @Synchronized
    fun writeTcpData(sessionId: Int, data: ByteArray) {
        writeTcpData(sessionId, data, data.size)
    }

    @Synchronized
    fun writeTcpData(sessionId: Int, data: ByteArray, length: Int) {
        val payloadLength = 1 + 4 + length
        val header = ByteArray(9)
        header.writeInt(0, payloadLength)
        header[4] = FullTunnelMessageType.TcpData.code.toByte()
        header.writeInt(5, sessionId)
        output.write(header)
        output.write(data, 0, length)
    }

    @Synchronized
    fun writeTcpDataCopying(sessionId: Int, data: ByteArray) {
        writeFrame(ByteArrayOutput().apply {
            writeByte(FullTunnelMessageType.TcpData.code)
            writeInt(sessionId)
            write(data)
        }.toByteArray())
    }

    @Synchronized
    fun writeCloseTcp(sessionId: Int) {
        writeFrame(ByteArrayOutput().apply {
            writeByte(FullTunnelMessageType.CloseTcp.code)
            writeInt(sessionId)
        }.toByteArray())
    }

    @Synchronized
    fun writeUdpData(host: String, port: Int, data: ByteArray) {
        val hostBytes = host.toByteArray(Charsets.UTF_8)
        writeFrame(ByteArrayOutput().apply {
            writeByte(FullTunnelMessageType.UdpData.code)
            writeShort(hostBytes.size)
            write(hostBytes)
            writeShort(port)
            writeInt(data.size)
            write(data)
        }.toByteArray())
    }

    fun read(): FullTunnelMessage? {
        val header = readExact(4) ?: return null
        val length = ByteBuffer.wrap(header).int
        require(length in 1..(1024 * 1024)) { "Invalid frame length: $length" }
        val payload = readExact(length) ?: return null
        var offset = 0
        val type = FullTunnelMessageType.fromCode(payload[offset++].toInt() and 0xff)

        return when (type) {
            FullTunnelMessageType.OpenTcp -> {
                val sessionId = payload.readInt(offset).also { offset += 4 }
                val hostLength = payload.readShort(offset).also { offset += 2 }
                val host = payload.decodeToString(offset, offset + hostLength)
                offset += hostLength
                val port = payload.readShort(offset)
                FullTunnelMessage(type, sessionId, host, port)
            }
            FullTunnelMessageType.TcpData,
            FullTunnelMessageType.CloseTcp,
            FullTunnelMessageType.Error -> {
                val sessionId = payload.readInt(offset).also { offset += 4 }
                FullTunnelMessage(type, sessionId, data = payload.copyOfRange(offset, payload.size))
            }
            FullTunnelMessageType.UdpData -> {
                val hostLength = payload.readShort(offset).also { offset += 2 }
                val host = payload.decodeToString(offset, offset + hostLength)
                offset += hostLength
                val port = payload.readShort(offset).also { offset += 2 }
                val dataLength = payload.readInt(offset).also { offset += 4 }
                FullTunnelMessage(type, host = host, port = port, data = payload.copyOfRange(offset, offset + dataLength))
            }
        }
    }

    private fun writeFrame(payload: ByteArray) {
        val header = ByteArray(4)
        header.writeInt(0, payload.size)
        output.write(header)
        output.write(payload)
        output.flush()
    }

    private fun readExact(length: Int): ByteArray? {
        val buffer = ByteArray(length)
        var offset = 0
        while (offset < length) {
            val read = input.read(buffer, offset, length - offset)
            if (read < 0) return null
            offset += read
        }
        return buffer
    }

    private fun ByteArray.readInt(offset: Int): Int = ByteBuffer.wrap(this, offset, 4).int
    private fun ByteArray.readShort(offset: Int): Int = ByteBuffer.wrap(this, offset, 2).short.toInt() and 0xffff
    private fun ByteArray.writeInt(offset: Int, value: Int) {
        this[offset] = (value ushr 24).toByte()
        this[offset + 1] = (value ushr 16).toByte()
        this[offset + 2] = (value ushr 8).toByte()
        this[offset + 3] = value.toByte()
    }
}

private class ByteArrayOutput {
    private val bytes = java.io.ByteArrayOutputStream()
    fun writeByte(value: Int) = bytes.write(value)
    fun writeInt(value: Int) = write(ByteBuffer.allocate(4).putInt(value).array())
    fun writeShort(value: Int) = write(ByteBuffer.allocate(2).putShort(value.toShort()).array())
    fun write(value: ByteArray) = bytes.write(value)
    fun toByteArray(): ByteArray = bytes.toByteArray()
}

File: usb-cellular-tether/windows/FullTunnelProtocol.cs
using System.Buffers.Binary;
using System.Text;

namespace UsbCellularTether.Windows;

internal enum FullTunnelMessageType : byte
{
    OpenTcp = 1,
    TcpData = 2,
    CloseTcp = 3,
    UdpData = 4,
    Error = 5,
}

internal static class FullTunnelProtocol
{
    public static async Task WriteOpenTcpAsync(Stream stream, int sessionId, string host, int port, CancellationToken cancellationToken)
    {
        var hostBytes = Encoding.UTF8.GetBytes(host);
        using var payload = new MemoryStream();
        payload.WriteByte((byte)FullTunnelMessageType.OpenTcp);
        WriteInt(payload, sessionId);
        WriteShort(payload, hostBytes.Length);
        payload.Write(hostBytes);
        WriteShort(payload, port);
        await WriteFrameAsync(stream, payload.ToArray(), cancellationToken);
    }

    public static Task WriteTcpDataAsync(Stream stream, int sessionId, ReadOnlyMemory<byte> data, CancellationToken cancellationToken) =>
        WriteSessionDataFastAsync(stream, FullTunnelMessageType.TcpData, sessionId, data, cancellationToken);

    public static Task WriteCloseTcpAsync(Stream stream, int sessionId, CancellationToken cancellationToken) =>
        WriteSessionDataFastAsync(stream, FullTunnelMessageType.CloseTcp, sessionId, ReadOnlyMemory<byte>.Empty, cancellationToken);

    public static async Task WriteUdpDataAsync(Stream stream, string host, int port, ReadOnlyMemory<byte> data, CancellationToken cancellationToken)
    {
        var hostBytes = Encoding.UTF8.GetBytes(host);
        using var payload = new MemoryStream();
        payload.WriteByte((byte)FullTunnelMessageType.UdpData);
        WriteShort(payload, hostBytes.Length);
        payload.Write(hostBytes);
        WriteShort(payload, port);
        WriteInt(payload, data.Length);
        payload.Write(data.Span);
        await WriteFrameAsync(stream, payload.ToArray(), cancellationToken);
    }

    public static async Task<FullTunnelMessage?> ReadAsync(Stream stream, CancellationToken cancellationToken)
    {
        var header = new byte[4];
        if (!await ReadExactAsync(stream, header, cancellationToken)) return null;
        var length = BinaryPrimitives.ReadInt32BigEndian(header);
        if (length <= 0 || length > 1024 * 1024) throw new InvalidOperationException($"Invalid tunnel frame length: {length}");
        var payload = new byte[length];
        if (!await ReadExactAsync(stream, payload, cancellationToken)) return null;

        var offset = 0;
        var type = (FullTunnelMessageType)payload[offset++];
        return type switch
        {
            FullTunnelMessageType.OpenTcp => throw new InvalidOperationException("Windows should not receive OpenTcp."),
            FullTunnelMessageType.TcpData => ReadSessionMessage(type, payload, ref offset),
            FullTunnelMessageType.CloseTcp => ReadSessionMessage(type, payload, ref offset),
            FullTunnelMessageType.UdpData => ReadUdpMessage(payload, ref offset),
            FullTunnelMessageType.Error => ReadSessionMessage(type, payload, ref offset),
            _ => throw new InvalidOperationException($"Unknown tunnel message type: {type}"),
        };
    }

    private static FullTunnelMessage ReadSessionMessage(FullTunnelMessageType type, byte[] payload, ref int offset)
    {
        var sessionId = ReadInt(payload, ref offset);
        var data = payload[offset..];
        return new FullTunnelMessage(type, sessionId, null, 0, data);
    }

    private static FullTunnelMessage ReadUdpMessage(byte[] payload, ref int offset)
    {
        var hostLength = ReadShort(payload, ref offset);
        var host = Encoding.UTF8.GetString(payload, offset, hostLength);
        offset += hostLength;
        var port = ReadShort(payload, ref offset);
        var dataLength = ReadInt(payload, ref offset);
        var data = payload.AsSpan(offset, dataLength).ToArray();
        return new FullTunnelMessage(FullTunnelMessageType.UdpData, 0, host, port, data);
    }

    private static async Task WriteSessionDataFastAsync(Stream stream, FullTunnelMessageType type, int sessionId, ReadOnlyMemory<byte> data, CancellationToken cancellationToken)
    {
        var header = new byte[9];
        BinaryPrimitives.WriteInt32BigEndian(header.AsSpan(0, 4), 1 + 4 + data.Length);
        header[4] = (byte)type;
        BinaryPrimitives.WriteInt32BigEndian(header.AsSpan(5, 4), sessionId);
        await stream.WriteAsync(header, cancellationToken);
        if (!data.IsEmpty)
        {
            await stream.WriteAsync(data, cancellationToken);
        }
    }

    private static async Task WriteFrameAsync(Stream stream, byte[] payload, CancellationToken cancellationToken)
    {
        var header = new byte[4];
        BinaryPrimitives.WriteInt32BigEndian(header, payload.Length);
        await stream.WriteAsync(header, cancellationToken);
        await stream.WriteAsync(payload, cancellationToken);
        await stream.FlushAsync(cancellationToken);
    }

    private static async Task<bool> ReadExactAsync(Stream stream, Memory<byte> buffer, CancellationToken cancellationToken)
    {
        var readTotal = 0;
        while (readTotal < buffer.Length)
        {
            var read = await stream.ReadAsync(buffer[readTotal..], cancellationToken);
            if (read == 0) return false;
            readTotal += read;
        }
        return true;
    }

    private static void WriteInt(Stream stream, int value)
    {
        Span<byte> bytes = stackalloc byte[4];
        BinaryPrimitives.WriteInt32BigEndian(bytes, value);
        stream.Write(bytes);
    }

    private static void WriteShort(Stream stream, int value)
    {
        Span<byte> bytes = stackalloc byte[2];
        BinaryPrimitives.WriteUInt16BigEndian(bytes, (ushort)value);
        stream.Write(bytes);
    }

    private static int ReadInt(byte[] bytes, ref int offset)
    {
        var value = BinaryPrimitives.ReadInt32BigEndian(bytes.AsSpan(offset, 4));
        offset += 4;
        return value;
    }

    private static int ReadShort(byte[] bytes, ref int offset)
    {
        var value = BinaryPrimitives.ReadUInt16BigEndian(bytes.AsSpan(offset, 2));
        offset += 2;
        return value;
    }
}

internal sealed record FullTunnelMessage(
    FullTunnelMessageType Type,
    int SessionId,
    string? Host,
    int Port,
    byte[] Data);
